File: scm-java/src/test/scheme/SchemeTests.java
package scheme;

import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.IOException;
import java.io.PrintStream;
import java.io.PrintWriter;
import java.io.PushbackReader;
import java.io.StringReader;
import java.io.StringWriter;
import java.nio.charset.StandardCharsets;
import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class SchemeTests {

    // ----------------------------------------------------------------
    // Core runner: test-group/=> format with snapshot/restore
    // ----------------------------------------------------------------

    private static int[] runCoreTest(String filename) throws IOException {
        int tests = 0;
        int failures = 0;
        var failureOutput = new StringBuilder();

        Scheme.strictImports = true;
        Scheme scheme = Scheme.forProgram();

        var inputStream = SchemeTests.class.getResourceAsStream(filename);
        var content = new String(inputStream.readAllBytes(), StandardCharsets.UTF_8);
        inputStream.close();

        var stream = new TextStream(new PushbackReader(new StringReader(content)), filename);
        var preambleForms = new java.util.ArrayList<Object>();
        boolean snapshotTaken = false;

        Object firstForm = scheme.read(stream);
        scheme.eval(null, firstForm);
        preambleForms.add(firstForm);

        while (true) {
            Object form = scheme.read(stream);
            if (Value.isEOFObject(form)) break;
            if (!Value.isPair(form)) continue;

            Pair pair = Value.asPair(form);
            if (!"test-group".equals(pair.car)) {
                scheme.eval(null, form);
                preambleForms.add(form);
                continue;
            }

            if (!snapshotTaken) {
                scheme.snapshot();
                snapshotTaken = true;
            }

            Object body = pair.cdr;
            while (body != Value.NIL) {
                Object subform = Value.asPair(body).car;
                body = Value.asPair(body).cdr;

                if (Value.isPair(subform) && "=>".equals(Value.asPair(subform).car)) {
                    SourcePos assertPos = Value.asPair(subform).pos;
                    Object assertParts = Value.asPair(subform).cdr;
                    Object exprForm = Value.asPair(assertParts).car;
                    Object expectedForm = Value.asPair(Value.asPair(assertParts).cdr).car;

                    try {
                        tests++;
                        String actual = Value.printRep(scheme.eval(exprForm));
                        Object expectedValue = scheme.eval(expectedForm);
                        String expected;
                        if (Value.isString(expectedValue)) {
                            String s = new String(Value.asString(expectedValue));
                            expected = s.startsWith("#<") ? s : Value.printRep(expectedValue);
                        } else {
                            expected = Value.printRep(expectedValue);
                        }
                        if (!actual.equals(expected)) {
                            failures++;
                            String loc = assertPos != null ? " at " + assertPos.filename + ":" + assertPos.line : "";
                            failureOutput.append("  FAIL").append(loc).append("\n");
                            failureOutput.append("    code:     ").append(Value.printRep(exprForm)).append("\n");
                            failureOutput.append("    expected: ").append(expected).append("\n");
                            failureOutput.append("    actual:   ").append(actual).append("\n");
                        }
                    } catch (Exception e) {
                        failures++;
                        String loc = assertPos != null ? " at " + assertPos.filename + ":" + assertPos.line : "";
                        failureOutput.append("  FAIL").append(loc).append("\n");
                        failureOutput.append("    code:      ").append(Value.printRep(exprForm)).append("\n");
                        failureOutput.append("    exception: ").append(e.getMessage()).append("\n");
                    }
                } else {
                    scheme.eval(subform);
                }
            }

            scheme.reset();
            for (Object pf : preambleForms) scheme.eval(null, pf);
        }

        if (failureOutput.length() > 0) System.out.print(failureOutput);
        return new int[]{tests, failures};
    }

    // ----------------------------------------------------------------
    // Tests runner: evaluate file, it reports on its own
    // ----------------------------------------------------------------

    private static int[] runTestsTest(String resourceName, String displayName) throws IOException {
        var inputStream = SchemeTests.class.getResourceAsStream(resourceName);
        var content = new String(inputStream.readAllBytes(), StandardCharsets.UTF_8);
        inputStream.close();

        var total = 0;
        var failed = 0;

        Scheme.strictImports = true;
        Scheme scheme = Scheme.forProgram();
        scheme.evalString(content, displayName);
        total = (int) (long) Value.asInteger(scheme.evalString("(last-run-total-tests)", "scheme-tests"));
        failed = (int) (long) Value.asInteger(scheme.evalString("(last-run-failed-tests)", "scheme-tests"));

        return new int[]{total, failed};
    }

    // ----------------------------------------------------------------
    // Failure runner: output-comparison tests
    // ----------------------------------------------------------------

    private static int[] runFailureTests(String platform) throws IOException {
        File failuresDir = new File("test/failures");
        if (!failuresDir.exists()) return new int[]{0, 0};

        int tests = 0;
        int failures = 0;

        String[] files = failuresDir.list();
        if (files == null) return new int[]{0, 0};
        Arrays.sort(files);

        for (String file : files) {
            if (!file.startsWith("test") || !file.endsWith(".scm")) continue;
            tests++;

            String baseName = file.substring(0, file.length() - 4);
            // Look for platform-specific expected, then generic
            File expectedFile = new File(failuresDir, baseName + ".expected-" + platform);
            if (!expectedFile.exists())
                expectedFile = new File(failuresDir, baseName + ".expected");
            if (!expectedFile.exists()) {
                System.out.println("  " + file + " SKIP (no .expected file)");
                continue;
            }

            String content = new String(java.nio.file.Files.readAllBytes(
                new File(failuresDir, file).toPath()), StandardCharsets.UTF_8);
            String expectedOutput = new String(java.nio.file.Files.readAllBytes(
                expectedFile.toPath()), StandardCharsets.UTF_8);

            // Capture stdout during test
            PrintStream originalOut = System.out;
            var capturedOutput = new ByteArrayOutputStream();
            System.setOut(new PrintStream(capturedOutput, true, StandardCharsets.UTF_8));

            try {
                Scheme.strictImports = true;
                Scheme scheme = Scheme.forProgram();
                try {
                    scheme.evalString(content, file);
                } catch (SchemeError e) {
                    e.printStackTrace();
                }
            } finally {
                System.setOut(originalOut);
            }

            String actual = normalizeOutput(capturedOutput.toString(StandardCharsets.UTF_8));
            String expected = normalizeOutput(expectedOutput);

            if (actual.equals(expected)) {
                System.out.println("  " + file + " OK");
            } else {
                failures++;
                System.out.println("  " + file + " FAIL");
                System.out.println("    expected: " + expected.replace("\n", "\n              "));
                System.out.println("    actual:   " + actual.replace("\n", "\n              "));
            }
        }

        return new int[]{tests, failures};
    }

    private static String normalizeOutput(String s) {
        String[] lines = s.replace("\r\n", "\n").replace("\r", "\n").split("\n", -1);
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < lines.length; i++) {
            if (i > 0) sb.append("\n");
            sb.append(lines[i].stripTrailing());
        }
        // Trim trailing newlines
        while (sb.length() > 0 && sb.charAt(sb.length() - 1) == '\n')
            sb.setLength(sb.length() - 1);
        return sb.toString();
    }

    // ----------------------------------------------------------------
    // Main: run categories in order, with command-line filtering
    // ----------------------------------------------------------------

    public static void main(String[] args) throws IOException {
        boolean success = true;
        int grandTotalTests = 0;
        int grandTotalFailures = 0;

        // Parse command line
        Set<String> requestedCategories = new HashSet<>();
        Set<String> requestedSuites = new HashSet<>();
        boolean runAll = args.length == 0;

        for (String arg : args) {
            if (arg.contains(":")) {
                requestedCategories.add(arg.split(":")[0]);
                requestedSuites.add(arg.split(":")[1]);
            } else if (arg.equals("core") || arg.equals("tests") || arg.equals("failures"))
                requestedCategories.add(arg);
            else
                requestedSuites.add(arg);
        }

        // === core ===
        if (runAll || requestedCategories.contains("core") || !requestedSuites.isEmpty()) {
            System.out.println("=== core ===");
            int catTests = 0, catFailures = 0;
            File coreDir = new File("test/core");
            if (coreDir.exists()) {
                String[] files = coreDir.list();
                Arrays.sort(files);
                for (String file : files) {
                    if (!file.endsWith(".scm")) continue;
                    if (!runAll && !requestedSuites.isEmpty() && !requestedSuites.contains(file)
                        && !requestedSuites.contains(file.replace(".scm", ""))) continue;
                    if (!runAll && requestedSuites.isEmpty() && !requestedCategories.contains("core")) continue;

                    System.out.print("  " + file + " ");
                    try {
                        int[] result = runCoreTest("/core/" + file);
                        catTests += result[0];
                        catFailures += result[1];
                        if (result[1] == 0)
                            System.out.println("OK, " + result[0] + " tests");
                        else {
                            System.out.println("FAIL, " + result[0] + " tests, " + result[1] + " failed");
                            success = false;
                        }
                    } catch (SchemeError e) {
                        System.out.println("ERROR");
                        e.printStackTrace(System.out);
                        success = false;
                    } catch (Exception e) {
                        System.out.println("ERROR: " + e.getClass().getSimpleName() + ": " + e.getMessage());
                        e.printStackTrace(System.out);
                        success = false;
                    }
                }
            }
            System.out.println("  core: " + catTests + " tests, " + catFailures + " failures");
            System.out.println();
            grandTotalTests += catTests;
            grandTotalFailures += catFailures;
        }

        // === tests ===
        if (runAll || requestedCategories.contains("tests") || !requestedSuites.isEmpty()) {
            System.out.println("=== tests ===");
            int catTests = 0, catFailures = 0;
            File testsDir = new File("test/tests");
            if (testsDir.exists()) {
                String[] files = testsDir.list();
                Arrays.sort(files);
                for (String file : files) {
                    if (!file.endsWith(".scm")) continue;
                    if (!runAll && !requestedSuites.isEmpty() && !requestedSuites.contains(file)
                        && !requestedSuites.contains(file.replace(".scm", ""))) continue;
                    if (!runAll && requestedSuites.isEmpty() && !requestedCategories.contains("tests")) continue;

                    try {
                        // Tests category always uses SRFI-64 runner
                        int[] result = runTestsTest("/tests/" + file, file);

                        catTests += result[0];
                        catFailures += result[1];
                        if (result[1] > 0) {
                            success = false;
                        }
                    } catch (SchemeError e) {
                        System.out.println("ERROR");
                        e.printStackTrace(System.out);
                        success = false;
                    } catch (Exception e) {
                        System.out.println("ERROR: " + e.getClass().getSimpleName() + ": " + e.getMessage());
                        e.printStackTrace(System.out);
                        success = false;
                    }
                }
            }
            System.out.println("  tests: " + catTests + " tests, " + catFailures + " failures");
            System.out.println();
            grandTotalTests += catTests;
            grandTotalFailures += catFailures;
        }

        // === failures ===
        if (runAll || requestedCategories.contains("failures")) {
            System.out.println("=== failures ===");
            int[] result = runFailureTests("java");
            grandTotalTests += result[0];
            grandTotalFailures += result[1];
            if (result[1] > 0) success = false;
            System.out.println("  failures: " + result[0] + " tests, " + result[1] + " failures");
            System.out.println();
        }

        // === total ===
        System.out.println("=== total ===");
        if (success)
            System.out.println(grandTotalTests + " tests, 0 failures");
        else
            System.out.println(grandTotalTests + " tests, " + grandTotalFailures + " failures");

        System.exit(success ? 0 : 1);
    }
}

File: scm-csharp/tests/SchemeTests.cs
using System.Reflection;
using System.Text;
using System.Text.RegularExpressions;
using scheme;

namespace schemetests;

public class SchemeTests
{
    // ----------------------------------------------------------------
    // Core runner: test-group/=> format with snapshot/restore
    // ----------------------------------------------------------------

    private static (int tests, int failures, string details) RunCoreTest(string resourceName, string displayName)
    {
        int tests = 0;
        int failures = 0;
        var failureOutput = new StringBuilder();
        Scheme.StrictImports = true;
        Scheme scheme = Scheme.ForProgram();

        var assembly = typeof(SchemeTests).Assembly;
        var content = new StreamReader(
            assembly.GetManifestResourceStream(resourceName)!
        ).ReadToEnd();

        using var stream = new TextStream(new StringReader(content), displayName);
        var preambleForms = new List<object>();
        bool snapshotTaken = false;

        var firstForm = scheme.Read(stream);
        scheme.Eval(null, firstForm);
        preambleForms.Add(firstForm);

        while (true)
        {
            var form = scheme.Read(stream);
            if (Value.IsEOFObject(form)) break;
            if (!Value.IsPair(form)) continue;

            var pair = Value.AsPair(form);
            if (!pair.car.Equals("test-group"))
            {
                scheme.Eval(null, form);
                preambleForms.Add(form);
                continue;
            }

            if (!snapshotTaken)
            {
                scheme.Snapshot();
                snapshotTaken = true;
            }

            var body = pair.cdr;
            while (body != Value.NIL)
            {
                var subform = Value.AsPair(body).car;
                body = Value.AsPair(body).cdr;

                if (Value.IsPair(subform) && Value.AsPair(subform).car.Equals("=>"))
                {
                    var assertPos = Value.AsPair(subform).pos;
                    var assertParts = Value.AsPair(subform).cdr;
                    var exprForm = Value.AsPair(assertParts).car;
                    var expectedForm = Value.AsPair(Value.AsPair(assertParts).cdr).car;

                    try
                    {
                        tests++;
                        var actual = Value.PrintRep(scheme.Eval(exprForm));
                        var expectedValue = scheme.Eval(expectedForm);
                        string expected;
                        if (Value.IsString(expectedValue))
                        {
                            var s = new string(Value.AsString(expectedValue));
                            expected = s.StartsWith("#<") ? s : Value.PrintRep(expectedValue);
                        }
                        else
                            expected = Value.PrintRep(expectedValue);

                        if (!actual.Equals(expected))
                        {
                            failures++;
                            var loc = assertPos != null ? " at " + assertPos.filename + ":" + assertPos.line : "";
                            failureOutput.AppendLine("  FAIL" + loc);
                            failureOutput.AppendLine("    code:     " + Value.PrintRep(exprForm));
                            failureOutput.AppendLine("    expected: " + expected);
                            failureOutput.AppendLine("    actual:   " + actual);
                        }
                    }
                    catch (Exception e)
                    {
                        failures++;
                        var loc = assertPos != null ? " at " + assertPos.filename + ":" + assertPos.line : "";
                        failureOutput.AppendLine("  FAIL" + loc);
                        failureOutput.AppendLine("    code:      " + Value.PrintRep(exprForm));
                        failureOutput.AppendLine("    exception: " + e.Message);
                    }
                }
                else
                {
                    scheme.Eval(subform);
                }
            }

            scheme.Reset();
            foreach (var pf in preambleForms) scheme.Eval(null, pf);
        }

        return (tests, failures, failureOutput.ToString());
    }

    // ----------------------------------------------------------------
    // Tests runner: evaluate file, it reports on its own
    // ----------------------------------------------------------------

    private static (int tests, int failures) RunTestsTest(string resourceName, string displayName)
    {
        var assembly = typeof(SchemeTests).Assembly;
        var content = new StreamReader(
            assembly.GetManifestResourceStream(resourceName)!
        ).ReadToEnd();

        var total = 0;
        var failed = 0;

        Scheme.StrictImports = true;
        Scheme scheme = Scheme.ForProgram();
        scheme.EvalString(content, resourceName);
        total = (int) Value.AsInteger(scheme.EvalString("(last-run-total-tests)", "scheme-tests"));
        failed = (int) Value.AsInteger(scheme.EvalString("(last-run-failed-tests)", "scheme-tests"));

        return (total, failed);
    }

    // ----------------------------------------------------------------
    // Failure runner: output-comparison tests
    // ----------------------------------------------------------------

    private static (int tests, int failures) RunFailureTests(string platform)
    {
        var assembly = typeof(SchemeTests).Assembly;
        string[] names = assembly.GetManifestResourceNames();
        int tests = 0;
        int failures = 0;

        var testFiles = names
            .Where(n => n.Contains(".failures.") && n.EndsWith(".scm"))
            .OrderBy(n => n)
            .ToArray();

        foreach (var resourceName in testFiles)
        {
            tests++;
            var testFileName = resourceName.Split('.').Reverse().Skip(1).First() + ".scm";
            var baseName = resourceName.Substring(0, resourceName.Length - 4); // strip .scm

            // Find expected output (platform-specific first, then generic)
            string? expectedResource = null;
            var platformSpecific = baseName + ".expected_" + platform;
            var generic = baseName + ".expected";

            if (names.Contains(platformSpecific))
                expectedResource = platformSpecific;
            else if (names.Contains(generic))
                expectedResource = generic;

            if (expectedResource == null)
            {
                Console.WriteLine("  " + testFileName + " SKIP (no .expected file)");
                continue;
            }

            var content = new StreamReader(assembly.GetManifestResourceStream(resourceName)!).ReadToEnd();
            var expectedOutput = new StreamReader(assembly.GetManifestResourceStream(expectedResource)!).ReadToEnd();

            // Capture stdout during test execution
            var originalOut = Console.Out;
            var capturedOutput = new StringWriter();
            Console.SetOut(capturedOutput);

            try
            {
                Scheme.StrictImports = true;
                Scheme scheme = Scheme.ForProgram();
                try
                {
                    scheme.EvalString(content, testFileName);
                }
                catch (SchemeError e)
                {
                    e.PrintStackTrace();
                }
            }
            finally
            {
                Console.SetOut(originalOut);
            }

            var actual = NormalizeOutput(capturedOutput.ToString());
            var expected = NormalizeOutput(expectedOutput);

            if (actual == expected)
            {
                Console.WriteLine("  " + testFileName + " OK");
            }
            else
            {
                failures++;
                Console.WriteLine("  " + testFileName + " FAIL");
                Console.WriteLine("    expected: " + expected.Replace("\n", "\n              "));
                Console.WriteLine("    actual:   " + actual.Replace("\n", "\n              "));
            }
        }

        return (tests, failures);
    }

    private static string NormalizeOutput(string s)
    {
        // Normalize line endings and trim trailing whitespace per line
        return string.Join("\n",
            s.Replace("\r\n", "\n").Replace("\r", "\n")
             .Split('\n')
             .Select(line => line.TrimEnd()))
            .TrimEnd('\n');
    }

    // ----------------------------------------------------------------
    // Main: run categories in order, with command-line filtering
    // ----------------------------------------------------------------

    public static int Main(string[] args)
    {
        var assembly = Assembly.GetExecutingAssembly();
        string[] allResources = assembly.GetManifestResourceNames();
        bool success = true;
        int grandTotalTests = 0;
        int grandTotalFailures = 0;

        // Parse command line: category names, or suite names, or "category:suite"
        var requestedCategories = new HashSet<string>();
        var requestedSuites = new HashSet<string>();
        bool runAll = args.Length == 0;

        foreach (var arg in args)
        {
            if (arg.Contains(':'))
            {
                requestedCategories.Add(arg.Split(':')[0]);
                requestedSuites.Add(arg.Split(':')[1]);
            }
            else if (arg == "core" || arg == "tests" || arg == "failures")
                requestedCategories.Add(arg);
            else
                requestedSuites.Add(arg);
        }

        bool ShouldRunCategory(string cat) =>
            runAll || requestedCategories.Contains(cat) || requestedSuites.Count > 0;

        bool ShouldRunSuite(string suiteName) =>
            runAll || requestedCategories.Count > 0 && requestedSuites.Count == 0 ||
            requestedSuites.Contains(suiteName) ||
            requestedSuites.Contains(Path.GetFileNameWithoutExtension(suiteName));

        // === core ===
        if (ShouldRunCategory("core"))
        {
            Console.WriteLine("=== core ===");
            int catTests = 0, catFailures = 0;

            var coreFiles = allResources
                .Where(n => n.Contains(".core.") && n.EndsWith(".scm"))
                .OrderBy(n => n);

            foreach (var resourceName in coreFiles)
            {
                var displayName = resourceName.Split('.').Reverse().Skip(1).First() + ".scm";
                if (!ShouldRunSuite(displayName)) continue;

                Console.Write("  " + displayName + " ");
                try
                {
                    var (t, f, details) = RunCoreTest(resourceName, displayName);
                    catTests += t;
                    catFailures += f;
                    if (f == 0)
                        Console.WriteLine("OK, " + t + " tests");
                    else
                    {
                        Console.WriteLine("FAIL, " + t + " tests, " + f + " failed");
                        Console.Write(details);
                        success = false;
                    }
                }
                catch (SchemeError e)
                {
                    Console.WriteLine("ERROR");
                    var sw = new StringWriter();
                    var prev = Console.Out;
                    Console.SetOut(sw);
                    e.PrintStackTrace();
                    Console.SetOut(prev);
                    Console.Write(sw.ToString());
                    success = false;
                }
                catch (Exception e)
                {
                    Console.WriteLine("ERROR: " + e.GetType().Name + ": " + e.Message);
                    Console.WriteLine(e.StackTrace);
                    success = false;
                }
            }
            Console.WriteLine("  core: " + catTests + " tests, " + catFailures + " failures");
            Console.WriteLine();
            grandTotalTests += catTests;
            grandTotalFailures += catFailures;
        }

        // === tests ===
        if (ShouldRunCategory("tests"))
        {
            Console.WriteLine("=== tests ===");
            int catTests = 0, catFailures = 0;

            var testFiles = allResources
                .Where(n => n.Contains(".tests.") && n.EndsWith(".scm"))
                .OrderBy(n => n);

            foreach (var resourceName in testFiles)
            {
                var displayName = resourceName.Split('.').Reverse().Skip(1).First() + ".scm";
                if (!ShouldRunSuite(displayName)) continue;

                try
                {
                    var (t, f) = RunTestsTest(resourceName, displayName);

                    catTests += t;
                    catFailures += f;
                    if (f > 0) success = false;
                }
                catch (SchemeError e)
                {
                    Console.WriteLine("ERROR");
                    var sw = new StringWriter();
                    var prev = Console.Out;
                    Console.SetOut(sw);
                    e.PrintStackTrace();
                    Console.SetOut(prev);
                    Console.Write(sw.ToString());
                    success = false;
                }
                catch (Exception e)
                {
                    Console.WriteLine("ERROR: " + e.GetType().Name + ": " + e.Message);
                    Console.WriteLine(e.StackTrace);
                    success = false;
                }
            }
            Console.WriteLine("  tests: " + catTests + " tests, " + catFailures + " failures");
            Console.WriteLine();
            grandTotalTests += catTests;
            grandTotalFailures += catFailures;
        }

        // === failures ===
        if (ShouldRunCategory("failures"))
        {
            Console.WriteLine("=== failures ===");
            try
            {
                var (t, f) = RunFailureTests("csharp");
                grandTotalTests += t;
                grandTotalFailures += f;
                if (f > 0) success = false;
                Console.WriteLine("  failures: " + t + " tests, " + f + " failures");
            }
            catch (Exception e)
            {
                Console.Error.WriteLine("  Failure tests error: " + e.Message);
                success = false;
            }
            Console.WriteLine();
        }

        // === total ===
        Console.WriteLine("=== total ===");
        if (success)
            Console.WriteLine(grandTotalTests + " tests, 0 failures");
        else
            Console.WriteLine(grandTotalTests + " tests, " + grandTotalFailures + " failures");

        return success ? 0 : 1;
    }
}
